src/main/java/io/nats/client/impl/NatsConnectionWriter.java

```diff
@@ -122,60 +123,65 @@ boolean isRunning() {
-                    }
-                }
+        int sendPosition = 0;
+        int sbl = sendBufferLength.get();
 
-                ByteArrayBuilder bab = msg.getProtocolBab();
-                int babLen = bab.length();
-                System.arraycopy(bab.internalArray(), 0, sendBuffer, sendPosition, babLen);
-                sendPosition += babLen;
+        while (msg != null) {
+            long size = msg.getSizeInBytes();
 
-                sendBuffer[sendPosition++] = CR;
-                sendBuffer[sendPosition++] = LF;
+            if (sendPosition + size > sbl) {
+                if (sendPosition > 0) {
+                    dataPort.write(sendBuffer, sendPosition);
+                    connection.getNatsStatistics().registerWrite(sendPosition);
+                    sendPosition = 0;
+                }
+                if (size > sbl) { // have to resize b/c can't fit 1 message
+                    sbl = bufferAllocSize((int) size, BUFFER_BLOCK_SIZE);
+                    sendBufferLength.set(sbl);
+                    sendBuffer = new byte[sbl];
+                }
+            }
 
-                if (!msg.isProtocol()) {
-                    sendPosition += msg.copyNotEmptyHeaders(sendPosition, sendBuffer);
+            ByteArrayBuilder bab = msg.getProtocolBab();
+            int babLen = bab.length();
+            System.arraycopy(bab.internalArray(), 0, sendBuffer, sendPosition, babLen);
+            sendPosition += babLen;
 
-                    byte[] bytes = msg.getData(); // guaranteed to not be null
-                    if (bytes.length > 0) {
-                        System.arraycopy(bytes, 0, sendBuffer, sendPosition, bytes.length);
-                        sendPosition += bytes.length;
-                    }
+            sendBuffer[sendPosition++] = CR;
+            sendBuffer[sendPosition++] = LF;
 
-                    sendBuffer[sendPosition++] = CR;
-                    sendBuffer[sendPosition++] = LF;
-                }
+            if (!msg.isProtocol()) {
+                sendPosition += msg.copyNotEmptyHeaders(sendPosition, sendBuffer);
 
-                stats.incrementOutMsgs();
-                stats.incrementOutBytes(size);
+                byte[] bytes = msg.getData(); // guaranteed to not be null
+                if (bytes.length > 0) {
+                    System.arraycopy(bytes, 0, sendBuffer, sendPosition, bytes.length);
+                    sendPosition += bytes.length;
+                }
 
-                msg = msg.next;
+                sendBuffer[sendPosition++] = CR;
+                sendBuffer[sendPosition++] = LF;
             }
 
-            dataPort.write(sendBuffer, sendPosition);
-            connection.getNatsStatistics().registerWrite(sendPosition);
+            stats.incrementOutMsgs();
+            stats.incrementOutBytes(size);
+
+            msg = msg.next;
         }
-        finally {
-            writerLock.unlock();
+
+        // no need to write if there are no bytes

```

Comment:
I added this. It makes sense that there could be times when there are no messages pending so why try to write or even call write. There is probably a lower level check in the write call, but since DataPort is an interface, it won't hurt